pandas/core/algorithms.py

```diff
@@ -1991,7 +1991,11 @@ def diff(arr, n: int, axis: int = 0, stacklevel=3):
 
     elif is_integer_dtype(dtype):
         # We have to cast in order to be able to hold np.nan
-        dtype = np.float64
+        # int8, int16 are incompatible with float64
```

Comment:
where exactly is the NotImplmementedError coming from. This *should* work as-is.

Reply:
Thought this change made sense because of this logic:

https://github.com/pandas-dev/pandas/blob/66a30bd007e964815e3a7f5c0f19065ccc1e0aed/pandas/_libs/algos.pyx#L1270-L1277

Was raising in that second branch

Reply:
grr this is super annoying. ok can you add this reference where you are making the code change itself.

Reply:
e.g. here

Reply:
done